### PR TITLE
Feature/manage tags

### PR DESCRIPTION
**Pull request by Pilopa**
## Proposed changes
Tags can now be added and deleted through the UI. Signatures of tags are now also displayed in the ui.
Tags are not automatically signed because it is not supported by JGit as of yet.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] PR is well described
- [x] Related issues linked to PR if existing and labels set
- [x] Target branch is not master (in most cases develop should bet the target of choice) 
- [x] Code does not conflict with target branch
- [x] New code is covered with unit tests
- [x] CHANGELOG.md updated
- [ ] Definition of Done's fulfilled: [DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/definition-of-done.md) // [UI DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/ui-dod.md)
- [x] Documentation updated (only necessary for new features or changed behaviour)

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  public TagCommandBuilder(Repository repository, TagCommand command) {
+    this.repository = repository;
+    this.command = command;
+    this.eventBus = ScmEventBus.getInstance();
+  }
+
+  public TagCreateCommandBuilder create() {
+    return new TagCreateCommandBuilder();
+  }
```
Comment (pfeuffer) [inline scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java:49]: This is one of the places where we try to keep useful JavaDocs for plugin devs

File: scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  public TagDeleteCommandBuilder delete() {
+    return new TagDeleteCommandBuilder();
+  }
```
Comment (pfeuffer) [inline scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java:53]: This is one of the places where we try to keep useful JavaDocs for plugin devs

File: scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      request.setRevision(revision);
+      return this;
+    }
+
+    public TagCreateCommandBuilder setName(String name) {
+      request.setName(name);
+      return this;
+    }
+
+    public Tag execute() throws IOException {
```
Comment (pfeuffer) [inline scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java:68]: This is one of the places where we try to keep useful JavaDocs for plugin devs

File: scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  public class TagDeleteCommandBuilder {
+    private TagDeleteRequest request = new TagDeleteRequest();
+
+    public TagDeleteCommandBuilder setName(String name) {
+      request.setName(name);
+      return this;
+    }
+
+    public void execute() throws IOException {
```
Comment (pfeuffer) [inline scm-core/src/main/java/sonia/scm/repository/api/TagCommandBuilder.java:83]: This is one of the places where we try to keep useful JavaDocs for plugin devs

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/GitUtil.java [hunk 5]
```diff
@@ -73,116 +77,120 @@
...
+  /**
+   * Field description
+   */
   private static final String REMOTE_REF = "refs/remote/scm/%s/%s";
 
-  /** Field description */
+  /**
+   * Field description
+   */
   private static final int TIMEOUT = 5;
```
Comment (pfeuffer) [inline scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/GitUtil.java:139]: If you format all these useless comments, you can better remove them

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/GitUtil.java [hunk 6]
```diff
@@ -73,116 +77,120 @@
...
           RevCommit c = getCommit(repository, revWalk, e.getValue());
 
-          if (c != null)
-          {
+          if (c != null) {
             tags.put(c.getId(), e.getKey());
-          }
-          else if (logger.isWarnEnabled())
-          {
+          } else if (logger.isWarnEnabled()) {
```
Comment (pfeuffer) [inline scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/GitUtil.java:189]:
```suggestion
          } else {
```

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+  public Tag create(TagCreateRequest request) {
+    try (Git git = new Git(context.open())) {
+      String revision = request.getRevision();
+
+      RevObject revObject;
+      Long tagTime;
+
+      if (Strings.isNullOrEmpty(revision)) {
+        throw new IllegalArgumentException("Revision is required");
+      }
```
Comment (pfeuffer) [inline scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java:87]: What about name?

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    this.eventBus = eventBus;
+  }
+
+  @Override
+  public Tag create(TagCreateRequest request) {
+    try (Git git = new Git(context.open())) {
+      String revision = request.getRevision();
+
+      RevObject revObject;
+      Long tagTime;
```
Comment (pfeuffer) [inline scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java:83]: Variables should be declared near their usage

File: scm-ui/ui-types/src/Tags.ts [hunk 9]
```diff
@@ -23,10 +23,12 @@
  */
 
 import { Links } from "./hal";
+import {Signature} from "./Signature";
```
Comment (pfeuffer) [inline scm-ui/ui-types/src/Tags.ts:26]: Your formatter is broken

File: scm-ui/ui-webapp/public/locales/en/repos.json [hunk 10]
```diff
@@ -103,12 +103,40 @@
...
   },
   "tag": {
     "name": "Name",
     "commit": "Commit",
-    "sources": "Sources"
+    "sources": "Sources",
+    "delete": {
+      "button": "Delete tag",
+      "subtitle": "Delete tag",
+      "description": "Deleted tag can not be restored.",
```
Comment (pfeuffer) [inline scm-ui/ui-webapp/public/locales/en/repos.json:132]:
```suggestion
      "description": "Deleted tags cannot be restored.",
```

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagRootResource.java [hunk 11]
```diff
@@ -98,6 +109,53 @@ public Response getAll(@PathParam("namespace") String namespace, @PathParam("nam
...
+      schema = @Schema(implementation = ErrorDto.class)
+    ))
+  @ApiResponse(
+    responseCode = "500",
+    description = "internal server error",
+    content = @Content(
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    ))
+  public Response create(@PathParam("namespace") String namespace, @PathParam("name") String name, @Valid TagRequestDto tagRequest) throws IOException {
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagRootResource.java:141]: We should add an `@Operations` annotation like [this one](https://github.com/scm-manager/scm-manager/blob/develop/scm-webapp/src/main/java/sonia/scm/api/v2/resources/BranchRootResource.java#L214) here

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagRootResource.java [hunk 12]
```diff
@@ -145,6 +203,47 @@ public Response get(@PathParam("namespace") String namespace, @PathParam("name")
...
+      if (tagExists(tagName, repositoryService)) {
+        repositoryService.getTagCommand().delete()
+          .setName(tagName)
+          .execute();
+      }
+
+      return Response.noContent().build();
+    } catch (IOException e) {
+      return Response.serverError().build();
+    }
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagRootResource.java:244]: `IOExceptions` should be handled by exception mappers elsewhere. It should be safe just to declare them as `throws`.

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagRootResource.java [hunk 13]
```diff
@@ -98,6 +109,53 @@ public Response getAll(@PathParam("namespace") String namespace, @PathParam("nam
...
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    ))
+  public Response create(@PathParam("namespace") String namespace, @PathParam("name") String name, @Valid TagRequestDto tagRequest) throws IOException {
+    NamespaceAndName namespaceAndName = new NamespaceAndName(namespace, name);
+    String revision = tagRequest.getRevision();
+    String tagName = tagRequest.getName();
+    try (RepositoryService repositoryService = serviceFactory.create(namespaceAndName)) {
+      if (tagExists(tagName, repositoryService)) {
+        throw alreadyExists(entity(Tag.class, tagName).in(Repository.class, namespace + "/" + name));
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagRootResource.java:147]:
```suggestion
        throw alreadyExists(entity(Tag.class, tagName).in(repositoryService.getRepository()));
```

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagToTagDtoMapper.java [hunk 14]
```diff
@@ -52,15 +54,21 @@
 
   @Mapping(target = "date", source = "date", qualifiedByName = "mapDate")
   @Mapping(target = "attributes", ignore = true) // We do not map HAL attributes
-  public abstract TagDto map(Tag tag, @Context NamespaceAndName namespaceAndName);
+  @Mapping(target = "signatures")
+  public abstract TagDto map(Tag tag, @Context NamespaceAndName namespaceAndName, @Context Repository repository);
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagToTagDtoMapper.java:58]: No need to pass both, namespace-and-name and repository.

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagToTagDtoMapper.java [hunk 15]
```diff
@@ -52,15 +54,21 @@
...
 
   @ObjectFactory
-  TagDto createDto(@Context NamespaceAndName namespaceAndName, Tag tag) {
+  TagDto createDto(@Context NamespaceAndName namespaceAndName, @Context Repository repository, Tag tag) {
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/TagToTagDtoMapper.java:61]: No need to pass both, namespace-and-name and repository.

**Review by pfeuffer — CHANGES_REQUESTED** [review]
From a functional point of view this really looks good. There are some issues in the code, nonetheless.

File: scm-it/src/test/java/sonia/scm/it/RepositoryAccessITCase.java [hunk 16]
```diff
@@ -175,12 +175,12 @@ public void shouldFindTags() throws IOException {
       .as("assert tag size")
       .isNotNull()
       .size()
-      .isGreaterThan(0);
+      .isPositive();
 
     assertThat(response.body().jsonPath().getMap("_embedded.tags.find{it.name=='" + tagName + "'}"))
       .as("assert tag has attributes for name, revision, date and links")
```
[the remark this answers is not in the document]
Comment (Pilopa) [reply scm-it/src/test/java/sonia/scm/it/RepositoryAccessITCase.java:181]: this was sonarqube :+1:

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+          .setObjectId(revObject)
+          .setTagger(new PersonIdent("SCM-Manager", "noreply@scm-manager.org"))
+          .setName(request.getName())
+          .call();
+
+      try (RevWalk walk = new RevWalk(git.getRepository())) {
+        revObject = walk.parseTag(ref.getObjectId());
+        final Optional<Signature> tagSignature = GitUtil.getTagSignature(revObject, gpg, walk);
+        tagSignature.ifPresent(tag::addSignature);
+      }
```
[the remark this answers is not in the document]
Comment (Pilopa) [reply scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java:112]: I do not understand the question unfortunately. We only add the signature if its present, which we dont know until we check for it.

[the remark this answers is not in the document]
Comment (Pilopa) [reply scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java:112]: Ah yeah this is future proofing once jgit implemented the feature :D

File: scm-plugins/scm-hg-plugin/src/test/java/sonia/scm/repository/spi/HgTagsCommandTest.java [hunk 18]
```diff
@@ -42,4 +61,15 @@ public void shouldGetTagDatesCorrectly() {
...
+  @Test
+  public void shouldNotDie() throws IOException {
+    HgTagCommand hgTagCommand = new HgTagCommand(cmdContext, workingCopyFactory);
+    new TagCommandBuilder(cmdContext.get(), hgTagCommand).create().setRevision("79b6baf49711").setName("newtag").execute();
+
+    HgTagsCommand hgTagsCommand = new HgTagsCommand(cmdContext);
+    final List<Tag> tags = hgTagsCommand.getTags();
+
+    assertThat(tags).isNotEmpty();
+  }
```
[the remark this answers is not in the document]
Comment (Pilopa) [reply scm-plugins/scm-hg-plugin/src/test/java/sonia/scm/repository/spi/HgTagsCommandTest.java]: Its for the bug that was fixed on a different pr ;)

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java (hunk 8 above)
Comment (Pilopa) [reply scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitTagCommand.java:83]: its used in multiple places and in logging when an exception is thrown

File: scm-ui/ui-webapp/src/repos/tags/components/TagRow.tsx [hunk 19]
```diff
@@ -24,35 +24,48 @@
...
+  let deleteButton;
+  if ((tag?._links?.delete as Link)?.href) {
+    deleteButton = (
+      <a className="level-item" onClick={() => onDelete(tag)}>
+        <span className="icon is-small">
+          <Icon name="trash" className="fas" title={t("tag.delete.button")} />
+        </span>
+      </a>
+    );
+  }
```
[the remark this answers is not in the document]
Comment (Pilopa) [reply scm-ui/ui-webapp/src/repos/tags/components/TagRow.tsx:56]: Branches dont have a spinner either, we should then add it to both or neither

**Pull request merged**